File: paper/src/main/java/com/github/aecsocket/minecommons/paper/serializers/ParticleEffectSerializer.java
package com.github.aecsocket.minecommons.paper.serializers;

import com.github.aecsocket.minecommons.core.vector.cartesian.Vector3;
import com.github.aecsocket.minecommons.paper.effect.PaperParticleEffect;

import org.bukkit.Particle;
import org.checkerframework.checker.nullness.qual.Nullable;
import org.spongepowered.configurate.ConfigurationNode;
import org.spongepowered.configurate.serialize.SerializationException;
import org.spongepowered.configurate.serialize.TypeSerializer;

import static com.github.aecsocket.minecommons.core.serializers.Serializers.require;

import java.lang.reflect.Type;

/**
 * Type serializer for a {@link PaperParticleEffect}.
 */
public class ParticleEffectSerializer implements TypeSerializer<PaperParticleEffect> {
    /** A singleton instance of this serializer. */
    public static final ParticleEffectSerializer INSTANCE = new ParticleEffectSerializer();

    private static final String
        NAME = "name",
        COUNT = "count",
        SIZE = "size",
        SPEED = "speed",
        DATA = "data";

    @Override
    public void serialize(Type type, @Nullable PaperParticleEffect obj, ConfigurationNode node) throws SerializationException {
        if (obj == null) node.set(null);
        else {
            node.node(NAME).set(obj.name());
            node.node(COUNT).set(obj.count());
            node.node(SIZE).set(obj.size());
            node.node(SPEED).set(obj.speed());
            if (obj.data() != null)
                node.node(DATA).set(obj.data());
        }
    }

    @Override
    public PaperParticleEffect deserialize(Type type, ConfigurationNode node) throws SerializationException {
        Particle particle = require(node.node(NAME), Particle.class);
        return new PaperParticleEffect(
            particle,
            node.node(COUNT).getDouble(0),
            node.node(SIZE).get(Vector3.class, Vector3.ZERO),
            node.node(SPEED).getDouble(0),
            particle.getDataType() == void.class ? null : node.node(DATA).get(particle.getDataType())
        );
    }
}

File: paper/src/main/kotlin/com/github/aecsocket/alexandria/paper/serializer/ParticleEffectSerializer.kt
package com.github.aecsocket.alexandria.paper.serializer

import com.github.aecsocket.alexandria.core.effect.ParticleEffect
import com.github.aecsocket.alexandria.core.extension.force
import com.github.aecsocket.alexandria.core.spatial.Vector3
import com.github.aecsocket.alexandria.paper.effect.particleByKey
import net.kyori.adventure.key.Key
import org.spongepowered.configurate.ConfigurationNode
import org.spongepowered.configurate.kotlin.extensions.get
import org.spongepowered.configurate.serialize.TypeSerializer
import java.lang.reflect.Type

private const val PARTICLE = "particle"
private const val COUNT = "count"
private const val SIZE = "size"
private const val SPEED = "speed"
private const val DATA = "data"

object ParticleEffectSerializer : TypeSerializer<ParticleEffect> {
    override fun serialize(type: Type, obj: ParticleEffect?, node: ConfigurationNode) {
        if (obj == null) node.set(null)
        else {
            node.node(PARTICLE).set(obj.particle)
            node.node(COUNT).set(obj.count)
            node.node(SIZE).set(obj.size)
            node.node(SPEED).set(obj.speed)
            obj.data?.let { node.node(DATA).set(it) }
        }
    }

    override fun deserialize(type: Type, node: ConfigurationNode): ParticleEffect {
        val particle = node.node(PARTICLE).force<Key>()
        val data = particleByKey(particle)?.let {
            when (val dataType = it.dataType) {
                Void::class.java -> null
                /* note: it's valid to have the data be null here
                   even if the particle is deserialized with no data,
                   and the particle requires data,
                   it can always be provided through code later */
                else -> node.node(DATA).get(dataType)
            }
        }
        return ParticleEffect(
            node.node(PARTICLE).force(),
            node.node(COUNT).get { 0.0 },
            node.node(SIZE).get { Vector3.Zero },
            node.node(SPEED).get { 0.0 },
            data
        )
    }
}
